doc/tools/docgram/README.md

```diff
@@ -0,0 +1,194 @@
+non-terminal.  The combined effect of these two is to replace all uses of
+`LEFTQMARK` to `"?"`.
+
+Here are the current operations.  They are likely to be refined as we learn
+what operations are most useful while we update the mlg files and documentation:
+
+Global edits
+------------
+
+DELETE - deletes the specified non-terminals anywhere in the grammar.
+Useful for removing non-terminals that only do lookahead that shouldn't be
+in the documentation.
+
+RENAME - each production specifies an (old name, new name) pair of
+non-terminals to rename.
+
+SPLICE - requests replacing all uses of the nonterminals anywhere in the
+grammar with its productions and removing the non-terminal.  (Doesn't work
+recursively; splicing for both `A: [ | B ]` and `B: [ | C ]` has to be
+done in 2 SPLICE operations.)
+
+EXPAND_LISTS - expands LIST0, LIST1, LIST* ... SEP and OPT constructs into
+new non-terminals
+
+Local edits
+-----------
+
+DELETE <production> - removes the specified production from the grammar
+
+EDIT <production> - modifies the specified production using the following tags
+that appear in the specified production:
+  USE_NT <name> LIST* - extracts LIST* as new nonterminal with the specified
+    new non-terminal name
+  ADD_OPT <grammar symbol> - looks for a production that matches the specified
+    production **without** <grammar_symbol>.  If found, both productions are
+    replaced with single production with `OPT <grammar_symbol>`
+
+The current version handles a single USE_NT or ADD_OPT per EDIT.
+
+REPLACE <oldprod>
+WITH <newprod> (2 sequential productions) - removes <oldprod> and
+inserts <newprod> in its place.
+
+(any other NT name) - adds a new production (and possibly a new nonterminal)
+to the grammar.
+
+.rst file updates
+-----------------
+
+doc_gram updates the .rst file when it sees the following 3 lines
+
+```
+.. insertgram <start> <end>
+
+.. productionlist:: XXX
+```
+
+The end of the existing `productionlist` is recognized by a blank line or a line
+with `:name:` or `:undocumented:`
+
+Other details
+-------------
+
+The output identifies productions from plugins that aren't automatically loaded with
+`(* XXX plugin *)` in grammar files and with `     (XXX plugin)` in productionlists.
+If desired, this mechanism could be extended to tag certain productions as deprecated,
+perhaps in conjunction with a coqpp change.
+
+My impression is that the order of the productions affects parsing.  If this is
+correct, we should consider how appropriate it is to document the grammar as if
+it is a context free grammar.
+
+In a number of places I've had to guess how the mlg files are interpreted since
+I've found no documentation.  It seems like there is some scoping of non-terminal
+names that I don't fully understand.  For example, there are 2 identical `scope`
+productions and 2 identical `orient` productions.  I was unable to figure out
+how to combine them.
```

Comment:
Should be understood and removed before merging.

Reply:
doc_grammar automatically and silently combines the identical productions.  Having these appear twice in the .mlgs is a minor wart to come back to later.  So I would leave this as-is.

There are still a number of details related to the .mlg files and parsing that I don't understand.  I had suggested to @ppedrot that it would help to have some documentation on this, but of course he has plenty to do.  We'll fix any resulting errors as we find them.

Reply:
Correction: I generate a warning message, such as:

```
Warning: *.mlg: Duplicate production 'record_fields: '
Warning: *.mlg: Duplicate production 'scope: "%" IDENT '
Warning: *.mlg: Duplicate production 'orient: "->" '
Warning: *.mlg: Duplicate production 'orient: "<-" '
Warning: *.mlg: Duplicate production 'orient: '
Warning: *.mlg: Duplicate production 'simple_tactic: "is_ground" constr      (* micromega plugin *) '
Warning: *.mlg: Duplicate production 'hloc: "in" "(" "Type" "of" ident ")"      (* ssr plugin *) '
Warning: *.mlg: Duplicate production 'hloc: "in" "(" "Value" "of" ident ")"      (* ssr plugin *) '
Warning: *.mlg: Duplicate production 'constr_eval: "type" "of" Constr.constr      (* ssr plugin *) '
```